### DISCUSSION TITLE
Compiled boot.py does not load

### DISCUSSION THREAD

**Discussion by cluuu** [micropython/micropython#10470] [ESP32] [answered]
Hello everyone, 

I just stumbled across the following issue: 
I compiled a bunch of `.py` to `.mpy` files using `mpy-cross` emitting mpy v6 according to the `--version` flag. After copying all of these files, some of my code -- specifically, the code in `boot.mpy` -- is not executed. I am using MicroPython v1.19.1 on an ESP32-C3 and have removed the `boot.py` file that was present after flashing the firmware onto the board. 

No code from my `boot.mpy` is executed. Is this expected behaviour? Putting the respective `boot.py` file onto the board whilst leaving everything else compiled files fixes the issue.

Is it possible to have `boot.mpy` as a compiled file? 

Thanks in advance.

**Comment by jimmo** [comment, marked as answer]
`boot.py` must be .py not .mpy. (Same goes for main.py too)

There should be very little code in boot.py, it should only be doing the bare minimum to set up filesystems so that main.py can be found, pre-compiling it will not make much difference.

main.py can then import pre-compiled files.

**Reply by cluuu** [reply]
Thanks for your reply, I guess this makes sense. 
I simply compiled all of the files I had written and was then wondering why things were not working.

Unfortunately, this does not seem to be mentioned in the documentation of `mpy` files. Do you think it would be beneficial to have that info in there? Would have at least saved me quite some time. I'd be happy to open a PR for this.

**Reply by jimmo** [reply]
Yep, that is a good idea. I am doing some work on the docs at the moment so will add it.

FWIW, there isn't really an obvious reason (to me at least) why we can't support compiled boot.py and main.py, so possibly this could also be fixed in the future.

**Comment by markusand** [comment]
Compiling boot.py may not make much difference, but compiling main.py may make it. I think it would be great to load main.mpy or fallback to main.py if it doesn't exist (or the reverse way).

At least, it would be great to have a flag in compilation to activate it.

**Comment by jcea** [comment]
My "main.py" is just a line: "import real_main", and that is a mpy ;-)

**Reply by jonnor** [reply]
This works rather well in practice. It can also be useful to only have watchdog + exception handler with reset in main.py, but keep it out of myapplication_main - since then one can do mpremote run myapplication_main.py and have it stop and raise the exception in REPL when debugging.

**Comment by jcea** [comment]
But, yes, boot and main should be able to be MPY too.

**Comment by sosi-deadeye** [comment]
For esp32 you can find the entry point here: https://github.com/micropython/micropython/blob/master/ports/esp32/main.c#L152C19-L152C40

The question is, how is the function called, which loads mpy files instead of py files.
